### PR TITLE
fix/ #2564 Cosmwasm pools in pool row should open in new tab

### PR DESCRIPTION
**Pull request by HarjobandeepSingh**
**What is the purpose of the change**
This pull request modifies the PaginatedTable component to open links in a new tab.

**Brief Changelog**
Modified the Link component in the PaginatedTable component to use a standard <a> tag with the target="_blank" attribute.

**Testing and Verifying**
This change has been tested locally by rebuilding the website and verifying that links in the PaginatedTable component open in a new tab as expected.

**Documentation and Release Note**
Does this pull request introduce a new feature or user-facing behavior changes? Yes
How is the feature or change documented? Not applicable

### CODE REVIEW

**Comment by HarjobandeepSingh** [comment]
fix/ #2564

**Comment by HarjobandeepSingh** [comment]
@MaxMillington Fix/

**Comment by CryptoAssassin1** [comment]
All the pools are opening in new tab.

File: packages/web/components/complex/paginated-table.tsx [hunk 1]
```diff
@@ -86,6 +86,7 @@ export const PaginatedTable = ({
               legacyBehavior
             >
               <a
+                target="_blank"
```
Comment (MaxMillington) [inline packages/web/components/complex/paginated-table.tsx:89]: this will make all pools open in a new tab, but only cosmwasm pools should open in a new tab since they link to celetone

**Review by MaxMillington — CHANGES_REQUESTED** [review]

**Comment by HarjobandeepSingh** [comment]
> Only cosmwasm pools should open in the tab currently all the pools are opening in new tab.
> 
> https://github.com/osmosis-labs/osmosis-frontend/assets/103904125/1027945a-0eb3-4934-baac-e39f1b103c07
> 
> 

I will fix it

**Comment by kamal-sutra** [comment]
Still non-cosmwasm pools also opening in new tab

**Comment by HarjobandeepSingh** [comment]
> Still non-cosmwasm pools also opening in new tab



> Still non-cosmwasm pools also opening in new tab. Please check again @HarjobandeepSingh

yea i acutally now i will push my final working my final code that is working

**Comment by HarjobandeepSingh** [comment]
@kamal-sutra plz check now

File: packages/web/components/complex/paginated-table.tsx [hunk 2]
```diff
@@ -79,25 +78,31 @@ export const PaginatedTable = ({
         {virtualRows.map((virtualRow) => {
           const row = rows[virtualRow.index] as Row<ObservablePoolWithMetric>;
           return (
-            <Link
```
Comment (MaxMillington) [inline packages/web/components/complex/paginated-table.tsx:82]: @HarjobandeepSingh why did you remove the `<Link />` component here? In general, we like to use the NextJS components unless there is a reason not to

File: packages/web/components/complex/paginated-table.tsx [hunk 3]
```diff
@@ -79,25 +78,31 @@ export const PaginatedTable = ({
...
+            <a
               key={row.original.queryPool.id}
               href={getPoolLink(row.original.queryPool)}
-              passHref
-              legacyBehavior
+              style={{
```
Comment (MaxMillington) [inline packages/web/components/complex/paginated-table.tsx:84]: we try to avoid inline styles, and instead use [tailwind](https://tailwindcss.com/) classes

File: packages/web/components/complex/paginated-table.tsx [hunk 4]
```diff
@@ -79,25 +78,31 @@ export const PaginatedTable = ({
...
+                position: "absolute",
+                top: 0,
+                left: 0,
+                width: "100%",
+                height: `${virtualRow.size}px`,
+                transform: `translateY(${virtualRow.start - topOffset}px)`,
+              }}
+              target={row.original.queryPool.type === "transmuter" ? "_blank" : "_self"}
```
Comment (MaxMillington) [inline packages/web/components/complex/paginated-table.tsx:92]: 👍

**Comment by kamal-sutra** [comment]
@HarjobandeepSingh Still non-cosmwasm pools opening in new tab.

**Comment by HarjobandeepSingh** [comment]
> @HarjobandeepSingh Still non-cosmwasm pools opening in new tab.

but my code working

**Comment by kamal-sutra** [comment]
https://github.com/osmosis-labs/osmosis-frontend/assets/116148606/cc985979-73c5-439f-8871-d14d4950417d

**Comment by HarjobandeepSingh** [comment]
> chrome_BsdYJlJQOU.mp4

can plz run this vercel test so i can check also online

File: packages/web/components/complex/paginated-table.tsx (hunk 2 above)
Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:82]: > @HarjobandeepSingh why did you remove the `<Link />` component here? In general, we like to use the NextJS components unless there is a reason not to

open some links in a new tab and others in the same tab. The Link component doesn't directly support this behavior

**Comment by arispech** [comment]
hmm I found that is not working again
Screenshot 2023-12-19 at 1 15 42 PM

**Comment by HarjobandeepSingh** [comment]
> > chrome_BsdYJlJQOU.mp4
> 
> can plz run this vercel test so i can check also online

@MaxMillington i again checked its working fine

**Comment by HarjobandeepSingh** [comment]
i test in local again it working

**Comment by arispech** [comment]
Can you see that it behaves differently? 
Round 1 pools that are opened in the same tap are 3,4,5,6,7
Round 2 pools that are opened in the new tap are 3,4,5,7

https://github.com/osmosis-labs/osmosis-frontend/assets/56182049/ef5f5686-74fd-4e3e-bb04-3db0c7eb20c3


https://github.com/osmosis-labs/osmosis-frontend/assets/56182049/57999d3a-7204-450f-be9f-1f56140d3e6a

**Comment by HarjobandeepSingh** [comment]
@arispech

**Comment by HarjobandeepSingh** [comment]
> @arispech 

 Plz checj now

**Review by CryptoAssassin1 — APPROVED** [review]
Worked fine but on mobile users will get this message.
image

**Comment by kamal-sutra** [comment]
@HarjobandeepSingh now cw pools doesn't open in new tab. both cw and normal pools open in same tab.
https://github.com/osmosis-labs/osmosis-frontend/assets/116148606/00e52244-a1c7-4c64-b572-82eafb68ad96

**Comment by HarjobandeepSingh** [comment]
> Worked fine but on mobile users will get this message. image

but this problem in that redirected site so what i can do ?

**Comment by CryptoAssassin1** [comment]
> @HarjobandeepSingh now cw pools doesn't open in new tab. both cw and normal pools open in same tab. https://github.com/osmosis-labs/osmosis-frontend/assets/116148606/00e52244-a1c7-4c64-b572-82eafb68ad96

@kamal-sutra cosmwasm pools are opening in new tab for me

**Comment by HarjobandeepSingh** [comment]
> > @HarjobandeepSingh now cw pools doesn't open in new tab. both cw and normal pools open in same tab. https://github.com/osmosis-labs/osmosis-frontend/assets/116148606/00e52244-a1c7-4c64-b572-82eafb68ad96
> 
> @kamal-sutra cosmwasm pools are opening in new tab for me

i think he is testing on wrong pools

**Comment by CryptoAssassin1** [comment]
> > Worked fine but on mobile users will get this message. image
> 
> but this problem in that redirected site so what i can do ?

@HarjobandeepSingh yes i know i left the comment so our team can see.

**Comment by kamal-sutra** [comment]
Oh sorry i thought to check cw20 token pools. Transmuter pools open in new tab. All good now.

**Review by kamal-sutra — APPROVED** [review]

File: packages/web/components/complex/paginated-table.tsx (hunk 2 above)
Comment (MaxMillington) [reply packages/web/components/complex/paginated-table.tsx:82]: @HarjobandeepSingh couldn't we do something like this work?

```
const handleLinkClick = (e, queryPool) => {
  e.stopPropagation();
  const link = getPoolLink(queryPool);

  if (queryPool.type === "transmuter") {
    window.open(link, '_blank');
  } else {
    router.push(link);
  }
};

// In your JSX
<Link href={getPoolLink(row.original.queryPool)} passHref>
  <a
    style={{
      position: "absolute",
      top: 0,
      left: 0,
      width: "100%",
      height: `${virtualRow.size}px`,
      transform: `translateY(${virtualRow.start - topOffset}px)`
    }}
    onClick={(e) => handleLinkClick(e, row.original.queryPool)}
  >
    {/* Link content */}
  </a>
</Link>
```

Removing the `<Link />` component is a performance regression since NextJS has built in pre-fetching with the `<Link />` component. So let's refactor to use `<Link />`

**Review by MaxMillington — CHANGES_REQUESTED** [review]
Let's use `<Link />`

File: packages/web/components/complex/paginated-table.tsx (hunk 2 above)
Comment (DavideSegullo) [reply packages/web/components/complex/paginated-table.tsx:82]: Are you sure about this behaviour? From next 13 "Link" component render a `<a/>` component:
https://nextjs.org/docs/pages/api-reference/components/link#props

https://nextjs.org/docs/pages/api-reference/components/link#legacybehavior

**Review by DavideSegullo — CHANGES_REQUESTED** [review]

**Comment by HarjobandeepSingh** [comment]
> > Let's use `<Link />`
> 
> > Let's use `<Link />`
> 
> but i think for this functionality using < a > tag would be better approach ?

**Review by arispech — APPROVED** [review]

File: packages/web/components/complex/paginated-table.tsx (hunk 2 above)
Comment (DavideSegullo) [reply packages/web/components/complex/paginated-table.tsx:82]: Why do you need `passHref` inside `<Link/>`? you can just add style to it (you also avoid one extra nest layer) and also if you use `target=_blank` on `<Link/>` maybe you don't need `onClick` event. Do you get any error if you don't use `onClick` event?

**Review by DavideSegullo — CHANGES_REQUESTED** [review]

File: packages/web/components/complex/paginated-table.tsx (hunk 2 above)
Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:82]: its working perfectly without any issues

Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:82]: The passHref inside <Link> is crucial for correctly passing the href property to the underlying <a> tag, ensuring proper link functionality. While adding styles directly might be an option, the passHref approach is cleaner and maintains the intended behavior of Next.js navigation. Additionally, using target="_blank" directly on <Link> could open all links in new tabs, affecting global behavior.

Comment (MaxMillington) [reply packages/web/components/complex/paginated-table.tsx:82]: https://nextjs.org/docs/pages/api-reference/components/link#prefetch

prefetch docs ^^^

Comment (DavideSegullo) [reply packages/web/components/complex/paginated-table.tsx:82]: You can specify `target="_self"` if query type isn't equal to `transmuter`

Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:82]: it might not have a significant impact on the results

**Comment by HarjobandeepSingh** [comment]
@MaxMillington sir , i have to change code again ?

**Comment by HarjobandeepSingh** [comment]
@DavideSegullo plz check now

File: packages/web/components/complex/paginated-table.tsx [hunk 5]
```diff
@@ -156,31 +164,29 @@ export const PaginatedTable = ({
...
-                      )}
-                    </Link>
-                  </td>
-                );
-              })}
-            </tr>
+          <tr
+          key={row.id}
+          className="transition-colors focus-within:bg-osmoverse-700 focus-within:outline-none hover:cursor-pointer hover:bg-osmoverse-800"
+          onClick={(e) => {
```
Comment (DavideSegullo) [inline packages/web/components/complex/paginated-table.tsx:170]: Again I don't like the use of the onClick event to handle the links, we lose the prefetch of the next links.
I suggest you look at how we implement this functionality in the asset table to use it as a reference:
https://github.com/osmosis-labs/osmosis-frontend/blob/stage/packages/web/components/table/index.tsx#L184

**Review by DavideSegullo — CHANGES_REQUESTED** [review]

**Comment by HarjobandeepSingh** [comment]
@DavideSegullo plz check now

File: packages/web/components/complex/paginated-table.tsx [hunk 6]
```diff
@@ -156,31 +159,31 @@ export const PaginatedTable = ({
...
-                    </Link>
-                  </td>
-                );
-              })}
-            </tr>
+<tr
+  key={row.id}
+  className="transition-colors focus-within:bg-osmoverse-700 focus-within:outline-none hover:cursor-pointer hover:bg-osmoverse-800"
+>
+  {row.getVisibleCells().map((cell) => {
```
Comment (DavideSegullo) [inline packages/web/components/complex/paginated-table.tsx:166]: why did you move the onClick event from the `tr` tag to the `td` tag?

File: packages/web/components/complex/paginated-table.tsx [hunk 7]
```diff
@@ -156,31 +159,31 @@ export const PaginatedTable = ({
...
+      <td
+        key={cell.id}
+        onClick={() => {
+          if (row.original.queryPool.type !== "transmuter") {
+            router.push(getPoolLink(row.original.queryPool));
+          }
+        }}
+      >
+        {row.original.queryPool.type === "transmuter" ? (
+          <a href={getPoolLink(row.original.queryPool)} target="_blank" rel="noopener noreferrer">
```
Comment (DavideSegullo) [inline packages/web/components/complex/paginated-table.tsx:177]: Why did you replace `Link` component with `a` tag?

File: packages/web/components/complex/paginated-table.tsx [hunk 8]
```diff
@@ -76,31 +76,34 @@ export const PaginatedTable = ({
...
-                <MobileTableRow row={row} />
-              </a>
-            </Link>
-          );
-        })}
+{virtualRows.map((virtualRow) => {
+  const row = rows[virtualRow.index] as Row<ObservablePoolWithMetric>;
+
+  return row.original.queryPool.type === "transmuter" ? (
+    <Link
```
Comment (DavideSegullo) [inline packages/web/components/complex/paginated-table.tsx:83]: Can't you just use a single `Link` component instead of two? You can use ternary operator inside `target` and `rel` props

**Review by DavideSegullo — CHANGES_REQUESTED** [review]

File: packages/web/components/complex/paginated-table.tsx (hunk 6 above)
Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:166]: The onClick event was moved from the tr tag to the td tag to allow for more granular control over the click events. This way, different actions can be performed based on which cell in the row was clicked. If the onClick event was on the tr tag, the same action would be performed no matter where in the row was clicked.

File: packages/web/components/complex/paginated-table.tsx (hunk 8 above)
Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:83]: Link dont support target so i also cant use <a> thats why did this

Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:83]: its working perfect still

File: packages/web/components/complex/paginated-table.tsx (hunk 7 above)
Comment (HarjobandeepSingh) [reply packages/web/components/complex/paginated-table.tsx:177]: The Link component was replaced with an a tag because the Link component from next/link does not support the target="_blank" attribute. This attribute is used to open the link in a new tab. The a tag was used instead to allow for this functionality.

**Comment by HarjobandeepSingh** [comment]
@DavideSegullo the code is perfectly working i dont think so it need changes

File: packages/web/components/complex/paginated-table.tsx (hunk 8 above)
Comment (MaxMillington) [reply packages/web/components/complex/paginated-table.tsx:83]: @HarjobandeepSingh I'm not sure I understand your comment, as you are using `target` as a prop here.

`<Link />` supports `target` as a prop in the latest version of NextJS https://nextjs.org/docs/app/api-reference/components/link#props

I agree with @DavideSegullo, it's easier to read if we just use one Link component and put the ternary in the `target` prop

**Comment by HarjobandeepSingh** [comment]
?

**Comment by MaxMillington** [comment]
@HarjobandeepSingh Thanks for the contribution! However, this issue has been fixed by another PR. Part of claiming the bounty is producing code that corresponds with our existing code style, and we review PRs in the order we receive them.

**Comment by HarjobandeepSingh** [comment]
> @HarjobandeepSingh Thanks for the contribution! However, this issue has been fixed by another PR. Part of claiming the bounty is producing code that corresponds with our existing code style, and we review PRs in the order we receive them.

i invested so much time in it 🥲 its ok
